src/hotspot/share/code/relocInfo.hpp

```diff
@@ -471,24 +473,81 @@ inline relocInfo prefix_relocInfo(int datalen = 0) {
 // the holder is "one size fits all".
 class RelocationHolder {
   friend class Relocation;
-  friend class CodeSection;
 
  private:
-  // this preallocated memory must accommodate all subclasses of Relocation
-  // (this number is assertion-checked in Relocation::operator new)
-  enum { _relocbuf_size = 5 };
-  void* _relocbuf[ _relocbuf_size ];
+  // A Relocation is "held" by placement constructing a Relocation into
+  // _relocbuf. Hence, _relocbuf must accomodate all subclasses of
+  // Relocation. We also need the Relocation base class to be at the same
+  // address as the start of the object, e.g. at the address of _relocbuf.
+  // Both of these requirements are checked in the constructors (see
+  // check_reloc_type and check_reloc_placement).  The placement of the base
+  // class subobject isn't guaranteed by C++, since these aren't standard
+  // layout classes, but all supported implementations provide that behavior.
+  // If that changes, we can instead add a Relocation* _reloc member to
+  // capture the result of the placement new, and use that to access the base
+  // subobject.
+  static const size_t _relocbuf_size = 5 * sizeof(void*);
+  union {
+    char _relocbuf[_relocbuf_size];
+    void* alignment_dummy;
+  };
+
+  template<typename Reloc>
+  void check_reloc_type() {
+    static_assert(std::is_base_of<Relocation, Reloc>::value, "not Relocation");
+    static_assert(sizeof(Reloc) <= sizeof(_relocbuf), "_relocbuf too small");
+  }
+
+  // Used to verify that the base class subobject of the object constructed
+  // into _relocbuf is at the same address as the derived object.
+  void check_reloc_placement(const Relocation* reloc) const {
+    assert(static_cast<const void*>(reloc) == _relocbuf, "invariant");
+  }
+
+  // Support for Relocation::copy_into.
+  // reloc should be a most derived object.
+  template<typename Reloc>
+  void copy_into_impl(const Reloc& reloc) {
+    check_reloc_type<Reloc>();
+    Relocation* copy = ::new (_relocbuf) Reloc(reloc);
+    check_reloc_placement(copy);
+  }
+
+  // Support for construct().
+  // ctor is a function object of one argument, the _relocbuf.  It should
+  // construct a relocation object in the given buffer and return a pointer to
+  // that new relocation object.
+  struct Construct {};          // Tag for selecting this constructor.
+  template<typename C> RelocationHolder(Construct, C ctor) {
+    using RelocPtr = decltype(ctor(_relocbuf));
+    static_assert(std::is_pointer<RelocPtr>::value, "expected relocation pointer");
+    using Reloc = std::remove_pointer_t<RelocPtr>;
+    check_reloc_type<Reloc>();
+    Relocation* reloc = ctor(_relocbuf);
+    check_reloc_placement(reloc);
+  }
```

Comment:
It seems that the current code is like this, rather than using `typename T, typename...Args` (where `T` is the relocation type), because it is not possible to specify explicit type arguments when calling a constructor template? (So you can't specify `T`).

But, you could truck in the type argument on `Construct`, and have it be inferred. So, this could become:

```suggestion
  template<typename T>
  struct Construct {};          // Tag for selecting this constructor.
  template<typename T, typename... Args> RelocationHolder(Construct<T>, const Args&...args) {
    check_reloc_type<T>();
    Relocation* reloc = ::new (_relocbuf) T(args...);
    check_reloc_placement(reloc);
  }
```
Callers provide e.g. `Construct<T>()` in `construct` or `Construct<Relocation>()` in the default constructor. `construct` can just forward the rest of the args.

This seems a little simpler (by avoiding the use of lambdas, and making it more locally obvious that the constructor is doing a placement new into `_relocbuf`), so maybe this is preferable?

Reply:
Yes, that works, and it's an improvement. Thanks. It also let me merge most of
the implementation of that constructor with the implementation of
`copy_into_impl`, which now use new `emplace_resource`.

I also took a moment to use `alignas` instead of a union, since JDK-8297912
was integrated earlier today.